### PR TITLE
Downstream QUIC/HTTP3 Support

### PR DESCRIPTION
**Pull request by tanujd11**
**What type of PR is this?**

Feature

**What this PR does / why we need it**:

Adds HTTP3 support in envoy gateway

**Which issue(s) this PR fixes**:

Fixes https://github.com/envoyproxy/gateway/issues/422

### CODE REVIEW

**Comment by arkodg** [comment]
thanks for building this out @tanujd11 ! , will start reviewing this post the v0.6 release

**Comment by arkodg** [comment]
hey @LanceEa can you help review this one ( since you have experience building this out before :) )

File: api/v1alpha1/clienttrafficpolicy_types.go
Comment (arkodg) [inline api/v1alpha1/clienttrafficpolicy_types.go:59]: lets make this `*bool` since its optional

File: internal/xds/translator/testdata/out/xds-ir/http3.routes.yaml
Comment (LanceEa) [inline internal/xds/translator/testdata/out/xds-ir/http3.routes.yaml:15]: This port needs to be the port exposed externally outside the cluster to the client. So, in most cases involving a browser it is more than likely going to be `:443` and not the listener port. The LoadBalancer of type `Service` then needs to point traffic to 10443. That is what I was referring to as the tricky part because a single Service doesn't always allow for MixedProtocol. _note: that might have change since last time I looked into this but it is the more nuanced part to the setup._

**Review by LanceEa — COMMENTED** [review]
I will defer to the maintainers on code coverage and general style but for the most part it looks good. I have left some feedback on the design and recommendations for future proofing it because support for mixedprotocols on the LoadBalancer Service and cloud vendors was limited last I looked.

File: internal/gatewayapi/listener.go
Comment (LanceEa) [reply internal/gatewayapi/listener.go:150]: This comment is referring to mixed protocols on a single Service of type `Loadbalancer`.

File: api/v1alpha1/clienttrafficpolicy_types.go
Comment (arkodg) [reply api/v1alpha1/clienttrafficpolicy_types.go:59]: I would love to just have a enable knob, but agree with @LanceEa, an empty struct maybe a good call for this API because some user may want to decide the listening port, and everyone loves tweaking port values to adhere with corp perimeter policies

**Comment by tanujd11** [comment]
Hi @LanceEa @arkodg I have made changes to the API making it as following
```
http3Settings:
  enabled: true
```
At the moment only enabled is there. I can raise a follow up PR to expose separate port for QUIC and other alt-svc configs. Till then this feature can be termed as experimental. WDYT?

**Comment by tanujd11** [comment]
Also it would be great if someone from @envoyproxy/gateway-reviewers / @envoyproxy/gateway-maintainers can try out the feature

File: api/v1alpha1/clienttrafficpolicy_types.go
Comment (arkodg) [inline api/v1alpha1/clienttrafficpolicy_types.go:70]: `Enabled` is not needed, instantiating the stunt should implicitly enable `http3`

File: api/v1alpha1/clienttrafficpolicy_types.go
Comment (arkodg) [inline api/v1alpha1/clienttrafficpolicy_types.go:64]: prefer `http3`

File: api/v1alpha1/clienttrafficpolicy_types.go
Comment (arkodg) [inline api/v1alpha1/clienttrafficpolicy_types.go:70]: Let's rm this TODO, this shows up on API docs

Comment (tanujd11) [reply api/v1alpha1/clienttrafficpolicy_types.go:70]: ohh shoot, completely slipped off my mind.

File: internal/gatewayapi/clienttrafficpolicy.go
Comment (arkodg) [inline internal/gatewayapi/clienttrafficpolicy.go:298]: nit: can we reuse `irListenerName` and also check if `irListenerName.Proxy != nil`

File: internal/ir/infra_test.go
Comment (arkodg) [inline internal/ir/infra_test.go:23]: why is this `true` ?

File: internal/infrastructure/kubernetes/proxy/resource_provider.go
Comment (arkodg) [inline internal/infrastructure/kubernetes/proxy/resource_provider.go:84]: should this live here on in the Gateway API layer  ?

File: internal/xds/translator/listener.go
Comment (arkodg) [inline internal/xds/translator/listener.go:167]: will auto work ? 
is there a way for us to support http3 but also support something else in case the client doesnt support it ?

File: internal/gatewayapi/testdata/clienttrafficpolicy-status-conditions.out.yaml
Comment (arkodg) [inline internal/gatewayapi/testdata/clienttrafficpolicy-status-conditions.out.yaml:285]: should address be made a pointer so its optional and doesnt show up in yaml ?

File: internal/xds/translator/listener.go
Comment (tanujd11) [reply internal/xds/translator/listener.go:167]: Hi Arko, If HTTP3 is not supported then the other listener will be used. We have created a mirror UDP listener and the connection would be upgraded to HTTP3 according to the alt-svc header. So even if HTTP3 is not supported HTTP/HTTP2 will be used

File: internal/infrastructure/kubernetes/proxy/resource_provider.go
Comment (tanujd11) [reply internal/infrastructure/kubernetes/proxy/resource_provider.go:84]: This is for opening UDP ports for only those ports which have HTTPS configured, That's a requirement for quic. So this condition is added here.

File: internal/ir/infra_test.go
Comment (tanujd11) [reply internal/ir/infra_test.go:23]: As we removed NewProxyListener by default and only add it when needed(Since we replaced only 1 listener supported to multiple listener), there is no error generated in this test case, hence false. Before any empty listener was there which failed the validate function.

**Comment by tanujd11** [comment]
Hi @arkodg , welcome back from the vacation. Your reviews on this will be greatly appreciated. Thanks.

**Review by arkodg — APPROVED** [review]
LGTM, thanks for adding this feature !

**Review by Xunzhuo — APPROVED** [review]
Great! Thanks for working on it : )

**Review by shawnh2 — APPROVED** [review]
lgtm

**Pull request merged**
